engine/exchange.go

```diff
@@ -59,78 +64,100 @@ func dryrunParamInteraction(param string) {
 	}
 }
 
-// GetExchangeByName returns an exchange given an exchange name
-func GetExchangeByName(exchName string) exchange.IBotExchange {
-	for x := range Bot.Exchanges {
-		if strings.EqualFold(Bot.Exchanges[x].GetName(), exchName) {
-			return Bot.Exchanges[x]
-		}
+func (e *exchangeManager) add(exch exchange.IBotExchange) {
+	e.m.Lock()
+	if e.exchanges == nil {
+		e.exchanges = make(map[string]exchange.IBotExchange)
 	}
-	return nil
+	e.exchanges[strings.ToLower(exch.GetName())] = exch
+	e.m.Unlock()
 }
 
-// ReloadExchange loads an exchange config by name
-func ReloadExchange(name string) error {
-	if len(Bot.Exchanges) == 0 {
-		return ErrNoExchangesLoaded
+func (e *exchangeManager) getExchanges() []exchange.IBotExchange {
+	if e.Len() == 0 {
+		return nil
 	}
 
-	e := GetExchangeByName(name)
-	if e == nil {
-		return ErrExchangeNotFound
+	e.m.Lock()
+	defer e.m.Unlock()
+	var exchs []exchange.IBotExchange
+	for x := range e.exchanges {
+		exchs = append(exchs, e.exchanges[x])
 	}
+	return exchs
+}
 
-	exchCfg, err := Bot.Config.GetExchangeConfig(name)
-	if err != nil {
-		return err
+func (e *exchangeManager) removeExchange(exchName string) error {
+	if e.Len() == 0 {
+		return ErrNoExchangesLoaded
 	}
-
-	e.Setup(exchCfg)
-	log.Debugf(log.ExchangeSys, "%s exchange reloaded successfully.\n", name)
+	exch := e.getExchangeByName(exchName)
+	if exch == nil {
+		return ErrExchangeNotFound
+	}
+	e.m.Lock()
+	defer e.m.Unlock()
+	delete(e.exchanges, strings.ToLower(exchName))
+	log.Infof(log.ExchangeSys, "%s exchange unloaded successfully.\n", exchName)
 	return nil
 }
 
-// UnloadExchange unloads an exchange by name
-func UnloadExchange(name string) error {
-	if len(Bot.Exchanges) == 0 {
-		return ErrNoExchangesLoaded
+func (e *exchangeManager) getExchangeByName(exchangeName string) exchange.IBotExchange {
+	if e.Len() == 0 {
+		return nil
 	}
-
-	if GetExchangeByName(name) == nil {
-		return ErrExchangeNotFound
+	e.m.Lock()
+	defer e.m.Unlock()
+	exch, ok := e.exchanges[strings.ToLower(exchangeName)]
+	if !ok {
+		return nil
 	}
+	return exch
+}
 
-	exchCfg, err := Bot.Config.GetExchangeConfig(name)
+func (e *exchangeManager) Len() int {
+	e.m.Lock()
+	defer e.m.Unlock()
+	return len(e.exchanges)
+}
+
+func (e *exchangeManager) unloadExchange(exchangeName string) error {
+	err := e.removeExchange(exchangeName)
 	if err != nil {
 		return err
 	}
 
-	exchCfg.Enabled = false
-	err = Bot.Config.UpdateExchangeConfig(exchCfg)
+	exchCfg, err := Bot.Config.GetExchangeConfig(exchangeName)
 	if err != nil {
 		return err
```

Comment:
Should the exchange be re-added here if this fails? A failed unloadExchange puts gct in a limbo where the exchange is not loaded, but the config says its enabled. Not that I'd expect this to fail, but more a precaution. Otherwise, this puts the onus on the calling function to determine the type of error and adjust accordingly

Reply:
That's a good point, I've changed the order of operations around so that both the config and removal have to pass before exchCfg.Enabled is set to false. Theoretically this should never be hit unless memory corruption occurs, since LoadExchange has to pass config checks also and add isn't exported